lib/internal/messages.js

```diff
@@ -0,0 +1,182 @@
+'use strict';
+
+/**
+ * Provides a simple internal API useful for externalizing Node.js' own
+ * output. The messages are defined as string templates compiled into
+ * the Node binary itself. The format is straightforward:
+ *
+ * const I18N = require('internal/messages');
+ *
+ * Message keys are integers defined as constants on the I18N object:
+ *   e.g. I18N.ERR_FOO, I18N.ERR_BAR, etc
+ *
+ * Given a hypothetical key ERR_FOO, with a message "Hello {0}",
+ *
+ *   const I18N = require('internal/messages');
+ *   console.log(I18N(I18N.ERR_FOO, 'World')); // Prints: Hello World
+ *
+ * If a particular message is expected to be used multiple times, the
+ * special `prepare()` method can be used to create a shortcut function:
+ *
+ *   const I18N = require('internal/messages');
+ *   const hello = I18N.prepare(I18N.ERR_FOO);
+ *   console.log(hello('World')); // Prints: Hello World
+ *
+ * As a convenience, it is possible to generate Error objects:
+ *
+ *    const I18N = require('internal/messages');
+ *    throw new I18N.TypeError('I18N.ERR_FOO, 'World');
+ *    throw new I18N.Error(I18N.ERR_FOO, 'World');
+ *    throw new I18N.RangeError(I18N.ERR_FOO, 'World');
+ *
+ * These Error objects will include both the expanded message and the
+ * string version of the message key in the Error.message. Each will
+ * include a `.key` property whose value is the string version of the
+ * message key, and a `.args` property whose values is an array of the
+ * replacement arguments passed in to generate the message.
+ *
+ * Calls to console.log() and console.error() can also be replaced with
+ * externalized versions:
+ *
+ *    const I18N = require('internal/messages');
+ *    I18N.log(I18N.ERR_FOO, 'World');
+ *      // Prints: `Hello World (ERR_FOO)` to stdout
+ *    I18N.error(I18N.ERR_FOO, 'World');
+ *      // Prints: `Hello World (ERR_FOO)` to stderr
+ *
+ **/
+
+const messages = process.binding('messages');
+
+function msg(key) {
+  return messages.messages[key];
+}
+
+function id(key) {
+  return messages.keys[key];
+}
+
+function replacement(msg, values) {
+  if (values && values.length > 0)
+    msg = msg.replace(/{(\d)}/g, (_, index) => values[Number(index)]);
+  return msg;
+}
+
+// I18N(key[, ...args])
+// Returns: string
+function message() {
+  const key = arguments[0];
```

Comment:
If the function always has one argument it's better to declare it as `function message(key) {`, that avoids an ArgumentsAdaptorTrampoline stack frame.